File: 프로그래머스/LV2/피로도.py
import itertools


def solution(k, dungeons):
    permutation_list = list(itertools.permutations(dungeons))

    answer = []
    for p in permutation_list:
        total = 0
        temp = k
        for i, j in p:
            if i > temp:
                break
            else:
                temp = temp - j
                total += 1

        answer.append(total)

    return max(answer)

File: 프로그래머스/javascript/LV2/피로도.js
function solution(k, dungeons) {
  let answer = 0;
  const visited = Array(dungeons.length).fill(false);

  function backtracking(hp, count) {
    answer = Math.max(answer, count);

    for (let i = 0; i < dungeons.length; i++) {
      const [minimum, need] = dungeons[i];
      if (!visited[i] && hp >= minimum) {
        visited[i] = true;
        backtracking(hp - need, count + 1);
        visited[i] = false;
      }
    }
  }

  backtracking(k, 0);
  return answer;
}
